pkg/workceptor/workceptor.go

```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 	return retMap, nil
 }
 
+func (w *Workceptor) cancelRemote(remoteNodeID, remoteUnitID string, unit *workUnit) error {
+	retryInterval := 500 * time.Millisecond
+	for {
+		// check that unit still exists, could have been released
+		unit.lock.RLock()
+		released := unit.released
+		unit.lock.RUnlock()
+		if released {
+			return nil
+		}
+		conn, _, err := w.connectToRemote(remoteNodeID)
+		if err != nil {
+			logger.Error("Connection failed to %s: %s\n", remoteNodeID, err)
+		}
+		if conn != nil {
+			_, err = conn.Write([]byte(fmt.Sprintf("work cancel %s\n", remoteUnitID)))
+			if err != nil {
+				logger.Error("Write error sending to %s: %s\n", remoteNodeID, err)
+			} else {
+				conn.Close()
+				break
+			}
+			conn.Close()
+		}
+		time.Sleep(retryInterval)
+		// backoff scheme, increase retryInterval
+		retryInterval = time.Duration(1.5 * float64(retryInterval))
+		if retryInterval > MaxWorkSleep {
+			retryInterval = MaxWorkSleep
+		}
+	}
+	retryInterval = 500 * time.Millisecond
+	var state int
+	for {
+		// check that unit still exists, could have been released
+		unit.lock.RLock()
+		released := unit.released
+		unit.lock.RUnlock()
+		if released {
+			return nil
+		}
+		unit.lock.RLock()
+		state = unit.status.State
+		unit.lock.RUnlock()
+		if IsComplete(state) {
+			return nil
+		}
+		time.Sleep(retryInterval)
+	}
+}
+
 // CancelUnit cancels a unit, killing it if it is still running
 func (w *Workceptor) CancelUnit(unitID string) error {
 	//TODO: Handle remote cancellation
 	w.scanForUnits()
 	w.activeUnitsLock.RLock()
 	unit, ok := w.activeUnits[unitID]
+	w.activeUnitsLock.RUnlock()
 	if !ok {
-		w.activeUnitsLock.RUnlock()
 		return fmt.Errorf("unknown work unit %s", unitID)
 	}
-	w.activeUnitsLock.RUnlock()
 	if unit.worker != nil {
 		err := unit.worker.Cancel()
 		if err != nil {
 			return err
 		}
 	}
-	unit.status.State = WorkStateFailed
-	unit.status.Detail = "Cancelled"
+	unit.lock.Lock()
+	unit.status.LocalCancel = true
+	nodeID := unit.status.Node
+	remoteUnitID := unit.status.RemoteUnitID
+	released := unit.released
+	unit.lock.Unlock()
+	if remoteUnitID != "" && !released {
+		unit.lock.Lock()
+		unit.status.Detail = "Cancel Pending"
+		unit.lock.Unlock()
+		go w.cancelRemote(nodeID, remoteUnitID, unit)
+	} else {
+		unit.lock.Lock()
+		unit.status.State = WorkStateFailed
+		unit.status.Detail = "Cancelled"
+		unit.lock.Unlock()
+	}
+	unitdir := path.Join(w.dataDir, unitID)
+	w.activeUnitsLock.RLock()
+	err := unit.status.Save(path.Join(unitdir, "status"))
+	w.activeUnitsLock.RUnlock()
+	if err != nil {
+		logger.Error("Error saving local status file: %s\n", err)
+		return nil
```

Comment:
Shouldn't we return the actual error here?

Reply:
fixed this